src/Storages/StorageS3.cpp

```diff
@@ -258,6 +258,7 @@ class StorageS3Source::DisclosedGlobIterator::Impl : WithContext
 
     KeyWithInfoPtr nextAssumeLocked()
     {
+        checkStackSize();
```

Comment:
Maybe just change recursion to `while` loop?

Reply:
ah yes, I just pushed the simplest fix without thinking more about it.
Thank for catching it!

Reply:
> AFAIU we can have deep recursion here when there are a lot of files in s3 and all of them don't match our globs pattern

That is correct.

> Maybe we can create a test with low value of setting s3_list_object_keys_size and large enough number of files

That's what I've been trying to do, the problem is that we have to create a lot of files which can take good amount of time, both from CH and directly into minio.
Other option is trying to create a mock server but it could get complex.

Reply:
Also with removed recursion we don't have anything to check for now, we would just set reading for arbitrary number of files which would always pass so the test would look useless to me IMO.